### ISSUE TITLE
use custom warnings instead of print statements (Trac #1428)

### ISSUE THREAD

**Issue by scipy-gitbot** [scipy/scipy#1953]
_Original ticket http://projects.scipy.org/scipy/ticket/1428 on 2011-04-28 by @josef-pkt, assigned to unknown._

use custom warnings instead of print statements 

eg. InterpolationWarning, OptimizationWarning, ....

```
From: Yaroslav Halchenko 

BUT scipy does use straight printouts of warnings


$> dpkg -L python-scipy | grep '\.py' | xargs grep -i "print.*warn" | head
/usr/share/pyshared/scipy/signal/signaltools.py:        print
"Warning: imaginary part of x ignored."
/usr/share/pyshared/scipy/signal/signaltools.py:        print
"Warning: imaginary part of x ignored."
/usr/share/pyshared/scipy/signal/filter_design.py:        print
"Warning, order is zero...check input parametegstop."
/usr/share/pyshared/scipy/interpolate/fitpack.py:                if
quiet<2:print 'Warning: Setting x[%d][%d]=x[%d][0]'%(i,m,i)
/usr/share/pyshared/scipy/interpolate/fitpack.py:            print
"Warning: "+_iermess[ier][0]
/usr/share/pyshared/scipy/interpolate/fitpack.py:            print
"Warning: "+_iermess[ier][0]
/usr/share/pyshared/scipy/interpolate/fitpack.py:            print
"Warning: the number of zeros exceeds mest"
/usr/share/pyshared/scipy/interpolate/fitpack.py:            print
"Warning: "+_iermess2[ierm][0]
/usr/share/pyshared/scipy/integrate/quadpack.py:            print
"Warning: " + msg
/usr/share/pyshared/scipy/maxentropy/maxentutils.py:#    print
"Warning: could not load the fast FORTRAN library for logsumexp()."


(git)novo:~/proj/scipy[tags/v0.9.0]git
$> git grep -i 'print .*Warning:'
scipy/integrate/quadpack.py:            print "Warning: " + msg
scipy/interpolate/fitpack.py:                if quiet<2:print
'Warning: Setting x[%d][%d]=x[%d][0]'%(i,m,i)
scipy/interpolate/fitpack.py:            print "Warning: "+_iermess[ier][0]
scipy/interpolate/fitpack.py:            print "Warning: "+_iermess[ier][0]
scipy/interpolate/fitpack.py:            print "Warning: the number of
zeros exceeds mest"
scipy/interpolate/fitpack.py:            print "Warning: "+_iermess2[ierm][0]
scipy/maxentropy/maxentutils.py:#    print "Warning: could not load
the fast FORTRAN library for logsumexp()."
scipy/maxentropy/maxentutils.py:        print "Warning: OverflowError
using numpy.exp(). Using slower Python"\
scipy/optimize/anneal.py:                print "Warning: Cooled to %f
at %s but this is not" \
scipy/optimize/anneal.py:            print "Warning: Maximum number of
iterations exceeded."
scipy/optimize/optimize.py:            print "Warning: Maximum number
of function evaluations has "\
scipy/optimize/optimize.py:            print "Warning: Maximum number
of iterations has been exceeded"
scipy/optimize/optimize.py:            print "Warning: Desired error
not necessarily achieved" \
scipy/optimize/optimize.py:            print "Warning: Maximum number
of iterations has been exceeded"
scipy/optimize/optimize.py:            print "Warning: Desired error
not necessarily achieved due to precision loss"
scipy/optimize/optimize.py:            print "Warning: Maximum number
of iterations has been exceeded"
scipy/optimize/optimize.py:            print "Warning: Maximum number
of iterations has been exceeded"
scipy/optimize/optimize.py:            print "Warning: Maximum number
of function evaluations has "\
scipy/optimize/optimize.py:            print "Warning: Maximum number
of iterations has been exceeded"
scipy/optimize/optimize.py:            print "Warning: Final
optimization did not succeed"
scipy/signal/signaltools.py:        print "Warning: imaginary part of
x ignored."
scipy/signal/signaltools.py:        print "Warning: imaginary part of
x ignored."
scipy/sparse/linalg/dsolve/umfpack/umfpack.py:
print 'warning: singular matrix'
scipy/sparse/linalg/dsolve/umfpack/umfpack.py:
print 'warning: recomputing symbolic'
scipy/sparse/linalg/dsolve/umfpack/umfpack.py:            print
'warning: (almost) singular matrix! '\
scipy/weave/blitz_tools.py:            print 'warning: compilation
failed. Executing as python code'
scipy/weave/build_tools.py:        #print "warning: build directory
was not part of python path."\
scipy/weave/build_tools.py:        print "warning: specified temp_dir
'%s' does not exist " \
scipy/weave/build_tools.py:            print "warning:, neither the
module's directory nor the "\
scipy/weave/build_tools.py:            print 'WARNING: failed to build
import library for gcc. Linking will fail.'
scipy/weave/catalog.py:        print 'warning: default directory is
not write accessible.'
scipy/weave/catalog.py:        print 'warning: default directory is
not write accessible.'
scipy/weave/catalog.py:            print 'warning: unable to repair
catalog entry\n %s\n in\n %s' % \
scipy/weave/swig2_spec.py:            print "WARNING: Multiple SWIG
versions detected.  No version was"
```

**Comment by ilayn** [comment]
This looks like invalid since quite some time now. Should we close it? 

ping @josef-pkt

**Comment by josef-pkt** [comment]
I don't know, I haven't paid enough attention to some implementation issues in scipy.

Does scipy now use mainly UserWarnings that warn only once, or customized warnings that always warn?

**Comment by rgommers** [comment]
This is still a thing:
```
$ rg "print\('Warning"
scipy/optimize/optimize.py
837:            print('Warning: ' + msg)
842:            print('Warning: ' + msg)

scipy/optimize/_trustregion.py
266:            print('Warning: ' + status_messages[warnflag])
(numpy-dev) rgommers@machine:~/code/scipy (master)$ rg "print\(\"Warning"
doc/sphinxext/numpydoc/docscrape.py
587:                print("Warning: invalid role %s" % self._role)

scipy/signal/filter_design.py
3674:        print("Warning, order is zero...check input parameters.")

scipy/optimize/optimize.py
3060:            print("Warning: " + msg)
3065:            print("Warning: " + msg)
3070:            print("Warning: " + msg)
3350:                print("Warning: Either final optimization did not succeed "

scipy/special/_basic.py
1168:        print("Warning, too many predicted coefficients.")
1225:        print("Warning, too many predicted coefficients.")
```

Warning can be replaced by `UserWarning`.

Getting rid of `disp` completely is out of scope for this issue:
```
        msg = _status_message['success']
        if disp:
            print(msg)
            print("         Current function value: %f" % fval)
            print("         Iterations: %d" % iterations)
            print("         Function evaluations: %d" % fcalls[0])
```

**Comment by dcolestock** [comment]
I'm looking into this issue as part of the pycon sprint

**Comment by StefanoFrazzetto** [comment]
Looking at ones in the `optimize` module

**Comment by AtsushiSakai** [comment]
https://github.com/scipy/scipy/pull/15259 fixed almost bare print statements, but the `_minimize_bfgs` ones are missed:
https://github.com/scipy/scipy/blob/2f36c582363b6e75ea1fca68bd87043f4d0b983c/scipy/optimize/_optimize.py#L1441-L1445
So, I reopened this. PR #14069 is trying to fix it.

**Comment by mdhaber** [comment]
gh-14069 merged, so I'll close this again.
